### DISCUSSION TITLE
Why does tree conditions stopped working

### DISCUSSION THREAD

**Discussion by DenisSvirin** [MilesCranmer/PySR#519] [Q&A] [answered]
Right now, I have custom_objective, where right part of a tree is divided into to parts: left (`tree_att`) and right (`power`).
I have constrains on `power` to be constant.

I need to calculate derivative of this tree, but I have to change it; to look like `tree_att ^ (2 * power)`.
If I do this like in the code below, trees in the hall of fame stop following constrains, to be particular **`power` is no longer a constant.**
Without changing a tree, there are no problems. 

**Why can this problem occur?**

```
    # Two body part
    E_two_body = tree

    E_two_body.degree != 2 && return L(Inf)

    # E repulsion
    tree_rep = E_two_body.l

    # E attraction, power
    tree_att_p = E_two_body.r

    tree_att_p.degree != 2 && return L(Inf)

    # E attraction
    tree_att = tree_att_p.l

    # power
    power = tree_att_p.r
    power.constant != true && return L(Inf)

    ...

    eval_power = power.val
    tree_att_p.op = 5 #options = [+, -, *, /, ^]
    tree_att_p.r.val = 2 .* abs(eval_power)
```

**Comment by MilesCranmer** [comment, marked as answer]
I think you need
```diff
-    power.constant != true && return L(Inf)
+    (power.degree != 0 || power.constant != true) && return L(Inf)
```

because the `.constant` field is some random value if it's an operator node. So for it to be a constant node it has to be `.degree == 0` AND `.constant == true`.

Also one other tip is that rather than returning `L(Inf)` for all of the constraints, you should return lesser and lesser penalties so that the evolutionary algorithm gets a sense of direction. e.g., for the first `L(Inf)`, make it `L(100_000)` (since that is the most important constraint to get right). The second `L(Inf)`, make it `L(50_000)`, since at least then it got the first constraint. Then `L(25_000)`. And so on. 

---

Also I would do

```julia
    eval_power = power.val::T
```

for improved speed

**Reply by DenisSvirin** [reply]
That works! Thanks
